# Выбор локации для скважины
____________

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Цель исследования</a></span></li><li><span><a href="#Задачи" data-toc-modified-id="Задачи-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Задачи</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Загрузка-и-подготовка-данных" data-toc-modified-id="Загрузка-и-подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка и подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Расчёт-прибыли-и-рисков" data-toc-modified-id="Расчёт-прибыли-и-рисков-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Расчёт прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

## Описание проекта

Компании «ГлавРосГосНефть» нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой Bootstrap.

Шаги для выбора локации:

* В избранном регионе собирают характеристики для скважин: качество нефти и объём её запасов;
* Строят модель для предсказания объёма запасов в новых скважинах;
* Выбирают скважины с самыми высокими оценками значений;
* Определяют регион с максимальной суммарной прибылью отобранных скважин

### Цель исследования

Построить модель машинного обучения для определения региона, где добыча принесёт наибольшую прибыль. Определить возможную прибыль и риски техникой Bootstrap.

### Задачи

1. Загрузить и подготовьть данные. 
2. Обучить и проверить модель для каждого региона:
 - Разбить данные на обучающую и валидационную выборки в соотношении 75:25.
 - Обучить модель и сделать предсказания на валидационной выборке.
 - Сохранить предсказания и правильные ответы на валидационной выборке.
 - Напечатать на экране средний запас предсказанного сырья и RMSE модели.
 - Проанализировать результаты.
3. Подготовить данные к расчёту прибыли:
 - Все ключевые значения для расчётов сохранить в отдельных переменных.
 - Рассчитать достаточный объём сырья для безубыточной разработки новой скважины. Сравнить полученный объём сырья со средним запасом в каждом регионе.
 - Написать выводы по этапу подготовки расчёта прибыли.
4. Написать функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели:
 - Выбрать скважины с максимальными значениями предсказаний.
 - Просуммировать целевое значение объёма сырья, соответствующее этим предсказаниям.
 - Рассчитать прибыль для полученного объёма сырья.
5. Посчитать риски и прибыль для каждого региона:
 -  Применить технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли.
 -  Найти среднюю прибыль, 95%-й доверительный интервал и риск убытков. Убыток — это отрицательная прибыль.
6. Написать выводы: предложить регион для разработки скважин и обосновать выбор.

### Описание данных

Вам предоставлены пробы нефти в трёх регионах. Характеристики для каждой скважины в регионе уже известны.

Данные геологоразведки трёх регионов находятся в файлах:
* `/datasets/geo_data_0.csv`;
* `/datasets/geo_data_1.csv`; 
* `/datasets/geo_data_2.csv`, где 

* `id` — уникальный идентификатор скважины;
* `f0`, `f1`, `f2` — три признака точек (неважно, что они означают, но сами признаки значимы);
* `product` — объём запасов в скважине (тыс. баррелей).

**Условия задачи:**

* Для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).
* При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
* Бюджет на разработку скважин в регионе — 10 млрд рублей.
* При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
* После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

## Загрузка и подготовка данных

Для начала импортируем необходимые для работы библиотеки.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from numpy.random import RandomState
import warnings
warnings.filterwarnings('ignore') 

pd.options.mode.chained_assignment = None 

Сохраним данные о трех скважинах в переменные `df1`, `df2` и `df3`. Выведем основные данные.

In [7]:
df1 = pd.read_csv('/datasets/geo_data_0.csv')
df2 = pd.read_csv('/datasets/geo_data_1.csv')
df3 = pd.read_csv('/datasets/geo_data_2.csv')

datasets = [df1, df2, df3]
_=0
for df in datasets:
    print('\033[31m', f'geo_data_{_}', '\033[0m')
    display(df.head())
    print(df.info(), '\n')
    _ += 1


 geo_data_0 


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None 

 geo_data_1 


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None 

 geo_data_2 


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None 



Во всех датасетах данные представлены в 5 столбцах. Пропусков нет.

Проверим дублируются ли данные о характеристиках скважин.

In [9]:
df_name = ['df1', 'df2', 'df3']
i=0

for df in datasets:
    print('\033[31m', df_name[i],'\033[0m', '\n', df.groupby('id')['id'].count().sort_values(ascending=False).head(11), '\n')
    i += 1

 df1  
 id
Tdehs    2
bxg6G    2
QcMuo    2
HZww2    2
TtcGQ    2
AGS9W    2
fiKDv    2
bsk9y    2
A5aEY    2
74z30    2
006OJ    1
Name: id, dtype: int64 

 df2  
 id
wt4Uk    2
LHZR0    2
5ltQ6    2
bfPNe    2
0022J    1
fDlc6    1
fDkrz    1
fDk0I    1
fDj3Z    1
fDiBm    1
fDebP    1
Name: id, dtype: int64 

 df3  
 id
xCHr8    2
KUPhW    2
Vcm5J    2
VF7Jo    2
fHfpS    1
fHeKc    1
fHeFi    1
fHYqC    1
fHY7n    1
fHX3Q    1
fHUkS    1
Name: id, dtype: int64 



Как видно, во всех трех регионах встречаются скважины, данные о которых записаны дважды. 

Это можно объяснить тем, исследования могли проводиться также дважды - в начале эксплуатации скважины и через какое-то время, когда представленные параметры в исходных данных изменились, так как в процессе нефтедобычи ухудшаются характеристики пласта и продукции вследствие повышения обводненности (образуются водонефтяные эмульсии, которые тяжелее добываются), выпадают соли, образуются асфальтеносмолопарафиновые отложения и т.д.

Выведем данные по нескольким таким скважинам.

In [5]:
display(df1[(df1['id'] == 'Tdehs') | (df1['id'] == 'AGS9W')].sort_values(by='id'))
display(df2[(df2['id'] == '5ltQ6') | (df2['id'] == 'wt4Uk')].sort_values(by='id'))
display(df3[(df3['id'] == 'xCHr8') | (df3['id'] == 'VF7Jo')].sort_values(by='id'))

id        f0        f1        f2     product
42529  AGS9W  1.454747 -0.479651  0.683380  126.370504
69163  AGS9W -0.933795  0.116194 -3.655896   19.230453
21426  Tdehs  0.829407  0.298807 -0.049563   96.035308
75715  Tdehs  0.112079  0.430296  3.218993   60.964018

id         f0         f1        f2     product
5849   5ltQ6  -3.435401 -12.296043  1.999796   57.085625
84461  5ltQ6  18.213839   2.191999  3.993869  107.813044
47591  wt4Uk  -9.091098  -8.109279 -0.002314    3.179103
82873  wt4Uk  10.259972  -9.376355  4.994297  134.766305

id        f0        f1        f2     product
11449  VF7Jo  2.122656 -0.858275  5.746001  181.716817
49564  VF7Jo -0.883115  0.560537  0.723601  136.233420
28039  xCHr8  1.633027  0.368135 -2.378367    6.120525
43233  xCHr8 -0.847066  2.101796  5.597130  184.388641

Как видно, сделанное предположение справедливо только для первого региона, так как в этом случае параметры `f0`, `f1`, `f2` изменились, а объемы запасов со временем уменьшились.

Для второго и третьего регионов объемы запасов увеличились со временем, что дает нам возможность предположить, что в столбце `product` представлены извлекаемые запасы нефти. Увеличение этого параметра может быть связано с применением методов повышения нефтеотдачи (химические обработки, тепловые методы и т.д.), что также влияет на пластовые свойства и свойства получаемой продукции (параметры `f0`, `f1`, `f2`) и, соответственно, увеличивает объемы возмжной для добычи нефти.

Так как в каждой "дублирующейся" скважине меняется не только параметр `product`, но и сами признаки, удалять их не потребуется - это дополнительные данные позволяющие получить зависимости между признаками и объемами добычи при обучении модели.

### Вывод

Исходные данные с информацией о признаках точек и объемах запасов для трех различных регионов были сохранены в переменные `df1`, `df2`, `df3`.

В каждом датасете представлено по 100 тыс. замеров характеристик `f0`, `f1` и `f2`и значений объема извлекаемых запасов нефти `product`.

Для некоторых скважин такие замеры делали дважды - в этом случае полученные характиристики отличались. Эти данные были оставлены без изменений для того, чтобы их также можно было использовать в дальнейшем анализе. 

## Обучение и проверка модели

Для каждого региона выделим общие признаки `f0`, `f1`, `f2` и запишем их в переменную features, целевой признак `product` сохраним в переменную target.

Данные разобьем на 2 выборки: обучающую и валидационную. Размеры обучающего и валидационного наборов определим в соотношении 75%:25%.

In [6]:
# фунуция по выделению общих и целевых признаков
def separation (df):
    target =  df['product']
    features = df.drop(['id', 'product'], axis=1)
    return target, features

# функция для разбивки данных на обучающую и валидационную выборки
def sample (features, target):
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target,test_size = 0.25, train_size =0.75, random_state=777)
    return features_train, features_valid, target_train, target_valid


target1, features1 = separation(df1)    
target2, features2 = separation(df2)  
target3, features3 = separation(df3)  

features_train1, features_valid1, target_train1, target_valid1 = sample(features1, target1)
features_train2, features_valid2, target_train2, target_valid2 = sample(features2, target2)
features_train3, features_valid3, target_train3, target_valid3 = sample(features3, target3)

print('Размеры train1:', features_train1.shape, 'Размеры valid1:', features_valid1.shape)
print('Размеры train2:', features_train2.shape, 'Размеры valid2:', features_valid2.shape)
print('Размеры train3:', features_train3.shape, 'Размеры valid3:', features_valid3.shape)


Размеры train1: (75000, 3) Размеры valid1: (25000, 3)
Размеры train2: (75000, 3) Размеры valid2: (25000, 3)
Размеры train3: (75000, 3) Размеры valid3: (25000, 3)


Обучим модель линейной регрессии и сделаем предсказания на валидационной выборке для каждого региона.

In [7]:
# функция по обучению модели
def model(features_train, target_train, features_valid, target_valid):
    model = LinearRegression()
    model.fit(features_train,target_train) 
    predictions_valid = model.predict(features_valid) 
    result = mean_squared_error(target_valid, predictions_valid)**0.5
    reserve = pd.Series(predictions_valid).mean()
    return predictions_valid, result, reserve


predictions_valid1, result1, reserve1 = model(features_train1, target_train1, features_valid1, target_valid1)
print('\033[1m', 'Регион 1', '\033[0m','\n','Средний запас продукции:', reserve1, '\n', 
      'RMSE модели линейной регрессии:', result1, '\n', sep='')

predictions_valid2, result2, reserve2 = model(features_train2, target_train2, features_valid2, target_valid2)
print('\033[1m', 'Регион 2', '\033[0m','\n','Средний запас продукции:', reserve2, '\n', 
      'RMSE модели линейной регрессии:', result2, '\n', sep='')

predictions_valid3, result3, reserve3 = model(features_train3, target_train3, features_valid3, target_valid3)
print('\033[1m', 'Регион 3', '\033[0m','\n','Средний запас продукции:', reserve3, '\n', 
      'RMSE модели линейной регрессии:', result3, '\n', sep='')


Регион 1
Средний запас продукции:92.71675294828383
RMSE модели линейной регрессии:37.99219303094202

Регион 2
Средний запас продукции:69.75032521418684
RMSE модели линейной регрессии:0.8911506041453005

Регион 3
Средний запас продукции:94.97788403611848
RMSE модели линейной регрессии:40.23018694848326



### Вывод

Получили, что по обученным моделям средний запас продукции на скважину максимален в 1 и 3 регионах: 92,72 и 94,98 тыс. баррелей соответственно. Меньшими средними запасами характиризуется 2 регион - всего 69,75 тыс. баррелей.

Однако меньшим различием между предсказаниями модели и историческими данными характеризуется именно 2 регион - RMSE составил всего 0,89. Для 1 и 3 регионов этот показатель составил 37,99 и 40,23 соответственно. Это говорит о том, что такую модель можно использовать в работе, но она будет не такой точной, как хотелось бы.

## Подготовка к расчёту прибыли

Определим достаточный объем сырья для безубыточной разработки новой скважины. 

Рассчитаем его как отношение бюджета на разработку скважин в регионе (10 млрд рублей) к произведению количества скважин в этом регионе, которые планируется пробурить, (200 шт.) и доходу с каждой единицы продукта(450 тыс. руб/ тыс. бар.), то есть:

${10*10^9\over 200*450*10^3}$ :




In [8]:
BUDGET = 10**10
WELLS_COUNT = 200
INCOME = 450*10**3

volume = BUDGET/WELLS_COUNT/INCOME
print('Достаточный объём сырья для безубыточной разработки новой скважины:', volume, 'тыс. баррелей')

Достаточный объём сырья для безубыточной разработки новой скважины: 111.11111111111111 тыс. баррелей


### Вывод

По итогам предварительных расчетов получили, что для того чтобы разработка 1 новой скважины являлась безубыточной, данная скважина должна обеспечивать объем добычи не менее 111.11 тыс. баррелей.

Ранее было получено, что средний запас продукции в 1 скважине для 1, 2 и 3 регионов составили соотвественно 92.72, 69.75 и 94.98 тысяч баррелей. Для первого и третьего регионов запасы достаточно близки к ннеобходимому количеству, но все же не дотягивают до него. 

Но этот показатель является средним и не говорит о том, в этих трех регионах нет скважин, способных дать необходимый объем продукции, поэтому необходимо будет также расчитать прибыль и возможные риски для каждого региона.

## Расчёт прибыли и рисков 

Напишем функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели, а также посчитаем риски и прибыль для каждого региона.

Для этого:
 - Выберем скважины с максимальными значениями предсказаний.
 - Просуммируем целевое значение объёма сырья, соответствующее этим предсказаниям.
 - Рассчитаем прибыль для полученного объёма сырья.
 - Применим технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли.
 - Найдем среднюю прибыль, 95%-й доверительный интервал и риск убытков. 


In [11]:
# функция по расчету прибыли
def revenue(target_valid, predictions_valid, count):
    predictions_sorted =  pd.Series(predictions_valid).sort_values(ascending=False)
    selected = target_valid[predictions_sorted.index][:count]
    profit = selected.sum() * INCOME - BUDGET
    return profit

# функция для расчета прибыли и рисков для топ-200 скважин
def rev_and_risk (target_valid, predictions_valid):
    values = []
    state = np.random.RandomState(12345)
    
    for i in range(1000):
        target_subsample = pd.Series(target_valid).reset_index(drop=True).sample(n=500, replace=True, random_state=state)
        predictions_subsample = pd.Series(predictions_valid)[target_subsample.index]
        values.append(revenue(target_subsample,predictions_subsample, WELLS_COUNT))
    
    values = pd.Series(values)
    lower = values.quantile(0.025)/10**6
    upper = values.quantile(0.975)/10**6
    
    risk = (values < 0).mean()*100
    val = values.mean()/10**6
    
    return val, lower, upper, risk

targets = [target_valid1, target_valid2, target_valid3]
predictions = [predictions_valid1, predictions_valid2, predictions_valid3]


for k in range(0,3):
    print('\033[1m', 'Регион ', k+1, '\033[0m') 
    val, lower, upper, risk = rev_and_risk(targets[k], predictions[k])
    print('Предполагаемая средняя прибыль с топ-200 скважин: ', '%.2f'%(val), ' млн. руб.','\n', 
     'Начало 95% доверительного интервала: ', '%.2f'%(lower), ' млн. руб., конец: ', '%.2f'%(upper), ' млн. руб.', '\n',
     'Риск убытков: ', '%.1f'%(risk), '%', '\n', sep = '')


 Регион  1 
Предполагаемая средняя прибыль с топ-200 скважин: 489.85 млн. руб.
Начало 95% доверительного интервала: -51.51 млн. руб., конец: 1046.14 млн. руб.
Риск убытков: 3.7%

 Регион  2 
Предполагаемая средняя прибыль с топ-200 скважин: 580.26 млн. руб.
Начало 95% доверительного интервала: 189.42 млн. руб., конец: 980.09 млн. руб.
Риск убытков: 0.2%

 Регион  3 
Предполагаемая средняя прибыль с топ-200 скважин: 393.94 млн. руб.
Начало 95% доверительного интервала: -136.53 млн. руб., конец: 958.35 млн. руб.
Риск убытков: 8.0%



### Вывод

В 1 регионе риск убытков составил 3.7%, во втором - 0.2%, в третьем - 8%.

Необходимо отобрать те регионы, в которых вероятность убытков меньше 2.5%. Этому условию удовлетворяет только 2 регион.
Соответствующая ему рассчитанная предположительная прибыль с топ-200 скважин составила 580.26 млн. руб.

## Общий вывод

В рамках произведенных расчетов было получено следующее:

* По предсказанию обученных моделей средние запасы продукции на скважину оказались максимальнны в 1 и 3 регионах: 92,72 и 94,98 тыс. баррелей соответственно. Меньшие средние запасы были получены для 2 региона - всего 69,75 тыс. баррелей. 

* Меньшим различием между предсказаниями модели и историческими данными характеризуется именно 2 регион - RMSE составил всего 0.89. Для 1 и 3 регионов этот показатель составил 37.99 и 40.23 соответственно. Это говорит о том, что такую модель можно использовать в работе, но она будет не такой точной, как хотелось бы.

* Для того чтобы разработка 1 новой скважины являлась безубыточной, данная скважина должна обеспечивать объем добычи не менее 111.11 тыс. баррелей.

* В 1 регионе риск убытков составил 3.7%, во втором - 0.2%, в третьем - 8%. Потенциальная прибыль - 489.85, 580.26, 393.94 млн. руб. соответственно.
_______________

При определении региона, где добыча принесёт наибольшую прибыль, отбирались те участки, для которых вероятность убытков составила меньше 2.5%. Этому условию удовлетворяет только 2 регион. Соответствующая ему рассчитанная предположительная прибыль с топ-200 скважин также оказалась максимальной среди рассмотренных трех регионов.

**Поэтому делаем вывод о том, что для разработки скважин наиболее перспективен 2 регион.**


